lib/module/src/module.rs

```diff
@@ -73,19 +94,56 @@ impl Linkage {
     }
 }
 
-#[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
-enum FuncOrDataId {
+
+/// A declared name may refer to either a function or data declaration
+#[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug)]
+pub enum FuncOrDataId {
+    /// When its a FuncId
     Func(FuncId),
+    /// When its a DataId
     Data(DataId),
 }
 
+/// Mapping to `ir::ExternalName` is trivial based on the `FuncId` and `DataId` mapping.
+impl From<FuncOrDataId> for ir::ExternalName {
+    fn from(id: FuncOrDataId) -> ir::ExternalName {
+        match id {
+            FuncOrDataId::Func(funcid) => ir::ExternalName::from(funcid),
+            FuncOrDataId::Data(dataid) => ir::ExternalName::from(dataid),
+        }
+    }
+}
+
 /// Information about a function which can be called.
 pub struct FunctionDeclaration {
     pub name: String,
     pub linkage: Linkage,
     pub signature: ir::Signature,
 }
 
+/// Error messages for all `Module` and `Backend` methods
+#[derive(Fail, Debug)]
+pub enum ModuleError {
+    /// Indicates an identifier was used before it was declared
+    #[fail(display = "Undeclared identifier {}", _0)]
+    Undeclared(String),
+    /// Indicates an identifier was used contrary to the way it was declared
+    #[fail(display = "Incompatible declaration of identifier {}", _0)]
+    IncompatibleDeclaration(String),
+    /// Indicates an identifier was defined more than once
+    #[fail(display = "Duplicate definition of identifier {}", _0)]
+    DuplicateDefinition(String),
+    /// Indicates an identifier was defined, but was declared as an import
+    #[fail(display = "Invalid definition of identifier {}", _0)]
+    InvalidDefinition(String),
```

Comment:
Can the error message and maybe also the enum arm for this one mention "imported", to be more specific about the problem? Something like `ImportGivenDefinition` and "Imported identifier assigned a definition"?